### PR TITLE
Remove identity-based collections from analysis

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

**Comment by findepi** [comment]
rebased on master to resolve conflicts.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalysis.java [hunk 1]
```diff
@@ -19,96 +19,102 @@
...
-        this.expressionTypes = requireNonNull(expressionTypes, "expressionTypes is null");
-        this.expressionCoercions = requireNonNull(expressionCoercions, "expressionCoercions is null");
-        this.typeOnlyCoercions = requireNonNull(typeOnlyCoercions, "typeOnlyCoercions is null");
-        this.columnReferences = new IdentityLinkedHashMap<>(requireNonNull(columnReferences, "columnReferences is null"));
-        this.subqueryInPredicates = requireNonNull(subqueryInPredicates, "subqueryInPredicates is null");
-        this.scalarSubqueries = requireNonNull(scalarSubqueries, "subqueryInPredicates is null");
-        this.existsSubqueries = requireNonNull(existsSubqueries, "existsSubqueries is null");
-        this.quantifiedComparisons = requireNonNull(quantifiedComparisons, "quantifiedComparisons is null");
-        this.lambdaArgumentReferences = requireNonNull(lambdaArgumentReferences, "lambdaArgumentReferences is null");
+        this.expressionTypes = new LinkedHashMap<>(requireNonNull(expressionTypes, "expressionTypes is null"));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalysis.java]: wrap with `unmodifiableSet` in constructor and then just return field reference from `get*` methods.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/NodeRefCollections.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    public static <T extends Node> Set<NodeRef<T>> fromIdentitySet(Set<T> set)
+    {
+        return set.stream()
+                .map(NodeRef::of)
+                .collect(toImmutableSet());
+    }
+
+    private NodeRefCollections()
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/NodeRefCollections.java]: 👎  since the class is removed at the end, fixuping it is wasteful :)

other that than, i'll try remember this advice

**Comment by findepi** [comment]
@losipiuk, comments applied.

rebased on master to resolve conflicts after #7712

**Comment by findepi** [comment]
@losipiuk i messed up with fixups, so they are squashed & rebased.

File: presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/db/TestResourceGroupsDao.java [hunk 3]
```diff
@@ -117,14 +117,12 @@ private static void testSelectorUpdate(H2ResourceGroupsDao dao, Map<Long, Select
                 2,
                 Optional.of(Pattern.compile("ping.*")),
                 Optional.of(Pattern.compile("ping_source")));
-        map.remove(2);
```
Comment (ArturGajowy) [inline presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/db/TestResourceGroupsDao.java]: unrelated?

Comment (ArturGajowy) [inline presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/db/TestResourceGroupsDao.java]: Also, the commit comment is hard to understand and needs rewording:
>Here fix is actually remove whenever .remove() with wrong type was
followed by .put() with a right one.

File: presto-main/src/main/java/com/facebook/presto/util/MoreMaps.java [hunk 4]
```diff
@@ -36,4 +39,18 @@ private MoreMaps() {}
                 .flatMap(Collection::stream)
                 .collect(toMap(Map.Entry::getKey, Map.Entry::getValue, merger));
     }
+
+    public static <K, V> Map<K, V> immutableLinkedMap(Map<K, V> map)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/MoreMaps.java]: Because it's not _linked_ (doesn't preserve order)? I'd be happy to drop this completely and use `ImmutableMap.copyOf` everywhere but I'd expect it'd be you, @ArturGajowy, who could object to that :D

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 5]
```diff
@@ -302,42 +302,31 @@ public Expression getJoinCriteria(Join join)
...
+        return ImmutableList.copyOf(existsSubqueries.get(NodeRef.of(node)));
     }
 
     public List<QuantifiedComparisonExpression> getQuantifiedComparisonSubqueries(Node node)
     {
-        if (quantifiedComparisonSubqueries.containsKey(node)) {
-            return quantifiedComparisonSubqueries.get(node);
-        }
-        return ImmutableList.of();
+        return unmodifiableList(quantifiedComparisonSubqueries.get(NodeRef.of(node)));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Because @sopel39 was worried about the cost of copying on access

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 6]
```diff
@@ -165,137 +165,139 @@ public void setCreateTableAsSelectNoOp(boolean createTableAsSelectNoOp)
...
     {
-        return orderByAggregates.get(query);
+        return orderByAggregates.get(NodeRef.of(node));
     }
 
-    public IdentityLinkedHashMap<Expression, Type> getTypes()
+    public Map<NodeRef<Expression>, Type> getTypes()
     {
-        return new IdentityLinkedHashMap<>(types);
+        return types;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Intentional. Look at other methods here -- the `Analysis` class has this peculiar habit of returning internal mutable state in getters. I do believe this should be FIXED but let's save this for another PR, if you may.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 7]
```diff
@@ -83,11 +84,10 @@
...
     private final Map<NodeRef<Join>, Expression> joins = new LinkedHashMap<>();
-    // TODO convert ListMultimap too
-    private final ListMultimap<Node, InPredicate> inPredicatesSubqueries = ArrayListMultimap.create();
-    private final ListMultimap<Node, SubqueryExpression> scalarSubqueries = ArrayListMultimap.create();
-    private final ListMultimap<Node, ExistsPredicate> existsSubqueries = ArrayListMultimap.create();
-    private final ListMultimap<Node, QuantifiedComparisonExpression> quantifiedComparisonSubqueries = ArrayListMultimap.create();
+    private final ListMultimap<NodeRef<Node>, InPredicate> inPredicatesSubqueries = ArrayListMultimap.create();
+    private final ListMultimap<NodeRef<Node>, SubqueryExpression> scalarSubqueries = ArrayListMultimap.create();
+    private final ListMultimap<NodeRef<Node>, ExistsPredicate> existsSubqueries = ArrayListMultimap.create();
+    private final ListMultimap<NodeRef<Node>, QuantifiedComparisonExpression> quantifiedComparisonSubqueries = ArrayListMultimap.create();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: it wasn't identity-based, but IMO it should have been

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 8]
```diff
@@ -303,10 +304,17 @@ public Expression getJoinCriteria(Join join)
...
-        this.scalarSubqueries.putAll(key, expressionAnalysis.getScalarSubqueries());
-        this.existsSubqueries.putAll(key, expressionAnalysis.getExistsSubqueries());
-        this.quantifiedComparisonSubqueries.putAll(key, expressionAnalysis.getQuantifiedComparisons());
+        this.inPredicatesSubqueries.putAll(key, dereference(expressionAnalysis.getSubqueryInPredicates()));
+        this.scalarSubqueries.putAll(key, dereference(expressionAnalysis.getScalarSubqueries()));
+        this.existsSubqueries.putAll(key, dereference(expressionAnalysis.getExistsSubqueries()));
+        this.quantifiedComparisonSubqueries.putAll(key, dereference(expressionAnalysis.getQuantifiedComparisons()));
+    }
+
+    private <T extends Node> List<T> dereference(Collection<NodeRef<T>> nodeRefs)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: It's not general purpose -- only here it was needed.

File: presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/db/TestResourceGroupsDao.java (hunk 3 above)
Comment (findepi) [inline presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/db/TestResourceGroupsDao.java]: Related: i was using intellin inspections to double-check i found every place to update, and this was showing up. It's separate commit, why not keep it in the PR? (I won't open another PR with just this, waste of time).

How can i improve the commit message?

File: presto-main/src/main/java/com/facebook/presto/util/MoreMaps.java (hunk 4 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/util/MoreMaps.java]: I double checked before asking that question, but now I see I've read the docs for `ImmutableMap.copyOf(Iterable)` - which says 'The returned map iterates over entries in the same order as the original iterable'. So, at the very least, we can `ImmutableMap.copyOf(map.entrySet())` in this method's body.

Also, the copyOf(Map) method's doc don't guarantee it to retain order, but it effectively does. So we can consider writing a learning test that will fail once they break the contract and use the copyOf.

> but I'd expect it'd be you, @ArturGajowy, who could object to that
not sure why though

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 8 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I see nothing non-general-purpose in its body or signature, and I expect to see more of its copies throughout the codebase soon enough to risk adding an (IMO) generic method only used once to the `NodeRef`. Not insisting though :)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 5 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: this itself doesn't explain why there's no worry for the other ones, but nvm

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 6 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: > save this for another PR
sure, but why break defensive copying in the places it's done? Especially for one of the seemingly widely used getter? Consistency is not a satisfactory answer :)

File: presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/db/TestResourceGroupsDao.java (hunk 3 above)
Comment (ArturGajowy) [inline presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/db/TestResourceGroupsDao.java]: I thought the `<Type>` addition was fixing the `map calls with wrong key type`. Now I see the `map.remove(2)` lines were doing the same.

Re commit message: I think a 'to' and 'the' are missing, and an object for 'remove'. So
>The fix is to remove the line whenever .remove() with wrong type is followed by .put() with the right one.

Comment (ArturGajowy) [inline presto-resource-group-managers/src/test/java/com/facebook/presto/resourceGroups/db/TestResourceGroupsDao.java]: Notice I had to read the commit again to invent the message, while the message should have helped me to understand the commit and its validity at the first reading

**Comment by findepi** [comment]
comments applied & rebased into master. @martint , in your hands now. This is many commits but hopefully quick read.

File: presto-main/src/main/java/com/facebook/presto/util/MoreMaps.java (hunk 4 above)
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/util/MoreMaps.java]: Immutable collections preserve the creation order. Neither of these helpers are needed and should be replaced with the immutable versions.

"Except for sorted collections, order is preserved from construction time."

https://github.com/google/guava/wiki/ImmutableCollectionsExplained

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/MoreMaps.java]: @electrum thanks for clarification! Commit dropped.

**Comment by findepi** [comment]
squashed & rebased

**Comment by findepi** [comment]
@martint, thanks for review!
